assets/js/wfparameters/Refine.js

```diff
@@ -0,0 +1,383 @@
+            }
+        }
+    }
+
+    loadHistogram(targetCol, baseState, clearEdits=false) {
+        // Loads a histogram from the server and sets the state with the result
+
+        // clearEdits controls whether we clear the edit on the server immediately upon load
+        // This is unused for now.
+        api.histogram(this.props.wfModuleId, targetCol)
+            .then(histogram => {
+                //console.log(histogram);
+                var nextState = Object.assign({}, baseState);
+                var editedHistogram = histogram.rows.map(function(entry) {
+                    var newEntry = Object.assign({}, entry);
+                    newEntry.selected = true;
+                    newEntry.edited = false;
+                    return newEntry;
+                });
+                //console.log(this.state.edits);
+                //console.log(editedHistogram);
+                // Apply all relevant edits we have to the original histogram
+                console.log(nextState.edits);
+                for(var i = 0; i < nextState.edits.length; i ++) {
+                    //console.log('applying edit');
+                    if(nextState.edits[i].column == this.props.selectedColumn) {
+                        //console.log('applying edit');
+                        editedHistogram = this.applySingleEdit(editedHistogram, nextState.edits[i]);
+                    }
+                }
+                //console.log(editedHistogram);
+                editedHistogram.sort((item1, item2) => {
+                    return item1.count < item2.count ? 1 : -1;
+                })
+                nextState.histogramData = editedHistogram;
+                nextState.histogramNumRows = editedHistogram.length;
+                nextState.histogramLoaded = true;
+                this.setState(nextState);
+                //console.log(nextState.histogramData);
+            })
+            .then(() => {
+                if(clearEdits) {
+                    this.props.saveEdits([]);
+                }
+            });
+    }
+
+    applySingleEdit(hist, edit) {
+        // Applies edits on the client side
+
+        //console.log(edit);
+        var newHist = hist.slice();
+        if(edit.type == 'change') {
+            var fromIdx = newHist.findIndex(function(element) {
+               return (element[edit.column] == edit.content.fromVal);
+            });
+            var fromEntry = Object.assign({}, newHist[fromIdx]);
+            newHist.splice(fromIdx, 1);
+            var toIdx = newHist.findIndex(function(element) {
+                return (element[edit.column] == edit.content.toVal);
+            });
+            if (toIdx == -1) {
+                // If no "to" entry was found, create a new entry
+                var newEntry = Object.assign({}, fromEntry);
+                newEntry[edit.column] = edit.content.toVal;
+                newEntry.edited = true;
+                newHist.unshift(newEntry);
```

Comment:
These four lines could be a single update() statement. See https://github.com/CJWorkbench/cjworkbench/blob/master/assets/js/workflow-reducer.js#L149